src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs

```diff
@@ -160,9 +164,13 @@ public static IResourceBuilder<ContainerResource> AddContainer(this IDistributed
     /// <returns></returns>
     public static IResourceBuilder<T> WithImageSHA256<T>(this IResourceBuilder<T> builder, string sha256) where T : ContainerResource
     {
-        var containerImageAnnotation = builder.Resource.Annotations.OfType<ContainerImageAnnotation>().Single();
-        containerImageAnnotation.SHA256 = sha256;
-        return builder;
+        if (builder.Resource.Annotations.OfType<ContainerImageAnnotation>().LastOrDefault() is { } existingImageAnnotation)
+        {
+            existingImageAnnotation.SHA256 = sha256;
+            return builder;
+        }
+
+        throw new InvalidOperationException($"The resource '{builder.Resource.Name}' does not have a container image specified. Use WithImage to specify the container image and tag.");
```

Comment:
(nit) maybe make a private helper for this exception so the message doesn't need to be duplicated (and doesn't get out of sync)?